feat: Multi-database and AWS S3 storage support

DevStarlight opened the pull request:
## Description

This PR introduces comprehensive database and storage flexibility to the Hyperfy server infrastructure. The changes add multi-database support (PostgreSQL, MySQL, SQLite) and AWS S3 storage capabilities, improving deployment flexibility and scalability options.

**Components affected:**
- Server infrastructure (database layer)
- Storage management system
- Asset handling and file storage
- World maintenance scripts

**Key improvements:**
- **Database flexibility**: Generic `DB_TYPE` configuration supporting PostgreSQL, MySQL, and SQLite
- **Storage options**: `STORAGE_TYPE` configuration for local file system or AWS S3 storage
- **Enhanced deployment**: Better support for cloud deployments and production environments
- **Improved maintenance**: Updated clean-world script supports all database types

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [x] New component
- [x] Component enhancement

## How Has This Been Tested?

Please describe your tests:

- [x] Component functionality tests
- [x] Integration tests with other Hyperfy components
- [x] World integration testing
- [ ] Cross-browser testing

**Test Configuration**:
* Hyperfy Version: 0.13.0
* Test world setup: Local development environment with SQLite fallback
* Browser(s): Chromium
* Node.js version: v22.13.1

## Checklist:

- [x] My code follows Hyperfy's component architecture
- [x] I have performed a self-review
- [x] I have documented the component's usage
- [x] I have tested the component in multiple scenarios
- [x] My changes maintain compatibility with existing worlds
- [x] I have updated the component documentation if needed

ashconnell committed:
Merge remote-tracking branch 'origin/dev'

ashconnell committed:
Merge remote-tracking branch 'origin/dev'

ashconnell committed:
Merge remote-tracking branch 'origin/dev'

ashconnell committed:
Merge remote-tracking branch 'origin/dev'

ashconnell committed:
Merge remote-tracking branch 'origin/dev'

ashconnell committed:
Merge remote-tracking branch 'origin/dev'

ashconnell committed:
Merge remote-tracking branch 'origin/dev'

ashconnell committed:
Merge remote-tracking branch 'origin/dev'

ashconnell committed:
Merge remote-tracking branch 'origin/dev'

DevStarlight committed:
feat: add multi-database support with generic environment variables

- Replace POSTGRES_* env vars with generic DB_* naming convention
- Add support for MySQL alongside PostgreSQL and SQLite
- Add DB_TYPE environment variable to specify database engine
- Configure optional dependencies for pg and mysql2 drivers
- Maintain backward compatibility with SQLite as fallback

DevStarlight committed:
feat: add multi-database support with generic environment variables

- Replace POSTGRES_* env vars with generic DB_* naming convention
- Add support for MySQL alongside PostgreSQL and SQLite
- Add DB_TYPE environment variable to specify database engine
- Configure optional dependencies for pg and mysql2 drivers
- Maintain backward compatibility with SQLite as fallback

DevStarlight committed:
feat: Add AWS S3 storage support with modular architecture

- Implement StorageManager with automatic S3/local detection
- Support S3 storage for assets, collections, and storage.json
- Add CloudFront integration and URL handling fixes
- Maintain full backward compatibility with local storage

DevStarlight committed:
refactor: Reorganize storage architecture and fix asset processing - Move storage files to dedicated directory, rename S3Storage to AwsS3Storage, remove redundant files, integrate CloudStorage into StorageManager, fix built-in collections asset processing, unify S3/local logic, improve collection listing format consistency, fix URL issues, add detailed logging

DevStarlight committed:
refactor: Reorganize storage architecture and fix asset processing - Move storage files to dedicated directory, rename S3Storage to AwsS3Storage, remove redundant files, integrate CloudStorage into StorageManager, fix built-in collections asset processing, unify S3/local logic, improve collection listing format consistency, fix URL issues, add detailed logging

DevStarlight committed:
Merge branch 'hyperfy-xyz:main' into feat/aws-storage-support

DevStarlight committed:
chore: Remove some useless comments

DevStarlight committed:
chore: Remove some useless comments

DevStarlight committed:
chore: Remove some useless methods

DevStarlight committed:
chore: Remove some useless methods

DevStarlight committed:
chore: include more needed venv variables

DevStarlight committed:
chore: include more needed venv variables

DevStarlight committed:
chore: Remove some extra lines

DevStarlight committed:
chore: Remove some extra lines

DevStarlight committed:
chore: Remove some extra lines

DevStarlight committed:
chore: Remove some extra lines

DevStarlight committed:
feat: add multi-database and AWS S3 storage support

- Add STORAGE_TYPE environment variable for explicit storage configuration
- Support local file system and AWS S3 storage with configurable switching
- Add multi-database support with PostgreSQL, MySQL, and SQLite fallback
- Update clean-world.mjs script to support all database types
- Improve environment variable documentation with clear configuration sections
- Add proper validation for required S3 configuration when using AWS storage
- Maintain backward compatibility with existing configurations

DevStarlight committed:
feat: modernize clean-world-s3 script with multi-database and STORAGE_TYPE support

DevStarlight committed:
feat: modernize clean-world-s3 script with multi-database and STORAGE_TYPE support

DevStarlight committed:
feat: Allow publishing branch from feat/full-storage-and-db-support

DevStarlight committed:
feat: Allow publishing branch from feat/full-storage-and-db-support

DevStarlight committed:
revert: remove docker.yml changes from feature branch

DevStarlight committed:
revert: remove docker.yml changes from feature branch

DevStarlight committed:
feat: Include support to define a schema in case there is only one database

DevStarlight committed:
feat: Include support to define a schema in case there is only one database

DevStarlight committed:
feat: Allow credentials to be defined as a service account

DevStarlight committed:
feat: Allow credentials to be defined as a service account

ashconnell reviewed (commented):
Just some quick observations, let me know your thoughts.

ashconnell commented:
Is it possible to rebase this branch so that it doesn't have a bunch of dev commits merged back on top of it? 

This helps a lot to keep git history clean

DevStarlight committed:
refactor(db): simplify DB config to only DB_TYPE and DB_URL, default to sqlite, remove MySQL support and old env vars. Update .env.example accordingly.

DevStarlight committed:
refactor(db): simplify DB config to only DB_TYPE and DB_URL, default to sqlite, remove MySQL support and old env vars. Update .env.example accordingly.

DevStarlight committed:
feat: Upgrade package.json & package-lock.json

DevStarlight commented on .env.example:
```diff
@@ -33,4 +33,55 @@ PUBLIC_ASSETS_URL=http://localhost:3000/assets
+S3_REGION=eu-west-1
+S3_ASSETS_PREFIX=
+S3_COLLECTIONS_PREFIX=
+S3_STORAGE_PREFIX=
```
I find your suggestion fine and I've updated both the code and the .env.example file to simplify the database setup as you recommended. Now, only two environment variables are used:

- DB_TYPE (leave empty for SQLite, or set to pg for Postgres)

- DB_URL (leave empty for SQLite, or provide a Postgres connection URI)

MySQL support and all the old related variables have been removed. This makes the default experience much simpler for new users, who can now spin up worlds with SQLite out of the box.

DevStarlight committed:
refactor(storage): rename aws storage to s3, update env vars and docs for S3-compatible storage support (CloudFlare R2, DigitalOcean Spaces, etc.)

DevStarlight committed:
refactor(storage): rename aws storage to s3, update env vars and docs for S3-compatible storage support (CloudFlare R2, DigitalOcean Spaces, etc.)

DevStarlight commented on scripts/clean-world-s3.mjs:
```diff
@@ -0,0 +1,269 @@
+
+// Close database connection before exiting
+await db.destroy()
+process.exit() 
```
You're right!

I've updated the codebase to use s3 instead of aws for the storage type and environment variables. This makes the storage integration more generic and compatible with any S3-compatible provider (such as CloudFlare R2, DigitalOcean Spaces, etc.), not just AWS S3.

All relevant code, environment variables, and documentation have been updated accordingly.

DevStarlight replied:
I've refactored the clean-world scripts to use shared utilities with functions like deleteFileS3() and deleteFileLocal().
Now both storage types stay in sync automatically when we update the cleanup logic - no more maintaining two separate scripts! The code is much cleaner and easier to maintain.

The scripts are organized in a nice modular structure (scripts/clean-world/) but still work exactly the same way with npm run world:clean.

Thanks for the feedback - this is definitely a much better approach!

DevStarlight commented on scripts/clean-world-s3.mjs:
```diff
@@ -0,0 +1,269 @@
+ * - For SQLite: Uses local ./[world]/db.sqlite file
+ * 
+ * Storage configuration:
+ * - STORAGE_TYPE must be set to 'aws'
```
You're right!

I've updated the codebase to use s3 instead of aws for the storage type and environment variables. This makes the storage integration more generic and compatible with any S3-compatible provider (such as CloudFlare R2, DigitalOcean Spaces, etc.), not just AWS S3.

All relevant code, environment variables, and documentation have been updated accordingly.

DevStarlight committed:
refactor(scripts): abstract clean-world scripts with shared utilities and modular structure

- Create clean-world/ directory with modular scripts
- Add clean-world-utils.mjs with shared functions (deleteFileLocal, deleteFileS3, etc.)
- Refactor clean-world-local.mjs and clean-world-s3.mjs to use shared utilities
- Update main clean-world.mjs to dynamically import appropriate script
- Add README.md with documentation for the new structure
- Fix path resolution issues for world directories
- Maintain backward compatibility with npm run world:clean

This addresses the reviewer's suggestion to avoid code duplication and ensure both storage types stay in sync when cleanup logic changes.

DevStarlight committed:
refactor(scripts): abstract clean-world scripts with shared utilities and modular structure

- Create clean-world/ directory with modular scripts
- Add clean-world-utils.mjs with shared functions (deleteFileLocal, deleteFileS3, etc.)
- Refactor clean-world-local.mjs and clean-world-s3.mjs to use shared utilities
- Update main clean-world.mjs to dynamically import appropriate script
- Add README.md with documentation for the new structure
- Fix path resolution issues for world directories
- Maintain backward compatibility with npm run world:clean

This addresses the reviewer's suggestion to avoid code duplication and ensure both storage types stay in sync when cleanup logic changes.

DevStarlight committed:
Merge branch 'feat/full-storage-and-db-support' of github.com:numengames/numinia-hyperfy2 into feat/full-storage-and-db-support

# Conflicts:
#	.env.example
#	scripts/clean-world.mjs
#	src/server/storage/StorageManager.js

DevStarlight committed:
fix: Solve an issue for what there was a double else block when initializing the storage manager

DevStarlight commented:
I did a rebase of the branch with dev and I think I didn't mess up anything 🤞!

ashconnell commented on package.json:
```diff
@@ -69,6 +69,12 @@
+    "pg": "^8.16.0",
+    "@aws-sdk/client-s3": "^3.685.0",
+    "@aws-sdk/s3-request-presigner": "^3.685.0",
+    "pg": "^8.16.0"
```
pg here twice

ashconnell commented on .env.example:
```diff
@@ -33,4 +33,49 @@ PUBLIC_ASSETS_URL=http://localhost:3000/assets
+S3_REGION=eu-west-1
+S3_ASSETS_PREFIX=
+S3_COLLECTIONS_PREFIX=
+S3_STORAGE_PREFIX=
```
can we do uri's for S3 too?

DevStarlight replied:
Implemented S3 URIs as suggested! 

Now you can use STORAGE_URL=s3://bucket-name/folder-path/assets/ with standard AWS credentials, (just like DB_URL approach).

ashconnell commented on scripts/clean-world.mjs:
```diff
@@ -1,153 +1,25 @@
+} else {
+    console.log('Running local cleanup...')
+    await import('./clean-world/clean-world-local.mjs')
+} 
```
we're still fragmenting here, eg any changes to the way cleaning works needs to be carefully updated in both of these and is likely to diverge because they are very different from each other. this is a risky piece of code.

i think this could be mitigated by having a single control flow that uses an abstraction to read and write from the db and files.

for example we have this code currently:

```jsx
let blueprints = new Set()
const blueprintRows = await db('blueprints')
for (const row of blueprintRows) {
  const blueprint = JSON.parse(row.data)
  blueprints.add(blueprint)
}
```

this would work with both database options.

but then we have this:

```jsx
console.log(`deleting ${filesToDelete.length} assets`)
for (const fileAsset of filesToDelete) {
  const fullPath = path.join(assetsDir, fileAsset)
  if (!DRY_RUN) {
    fs.removeSync(fullPath)
  }
  console.log('delete asset:', fileAsset)
}
```

which only works with the filesystem, but could be rewritten to work with both (this is an example only):

```jsx
console.log(`deleting ${filesToDelete.length} assets`)
for (const fileAsset of filesToDelete) {
  if (!DRY_RUN) {
    await storage.deleteAsset(fileAsset)
  }
  console.log('delete asset:', fileAsset)
}
```

this would allow us to have a SINGLE script for world cleaning instead of needing to keep two in sync and risk them diverging/breaking

DevStarlight committed:
Merge upstream/dev: Integrate latest changes with flexible DB and S3 storage support

- Resolved conflicts in clean-world.mjs: unified script supports both local and S3 storage
- Resolved conflicts in db.js: flexible database configuration (SQLite default, PostgreSQL via env vars)
- Resolved conflicts in index.js: integrated StorageManager for multi-storage support
- Added S3 storage capabilities via StorageManager architecture
- Maintained backward compatibility with local filesystem storage
- All database operations now support both SQLite and PostgreSQL

DevStarlight committed:
chore: Remove some useless imports in clean-world.mjs

DevStarlight committed:
cleanup: Remove obsolete clean-world fragments

- Deleted scripts/clean-world/clean-world-local.mjs (functionality moved to unified script)
- Deleted scripts/clean-world/clean-world-s3.mjs (functionality moved to unified script)
- Deleted scripts/clean-world/clean-world-utils.mjs (functions now inline)
- Deleted scripts/clean-world/README.md (documentation for removed fragmented approach)
- Removed empty scripts/clean-world/ directory

All functionality is now consolidated in scripts/clean-world.mjs with conditional logic for local/S3 storage.

DevStarlight committed:
fix: Remove duplicated lib imported in package.json

DevStarlight committed:
feat: Add S3 URI configuration support

- Added S3 URI parsing (s3://bucket/prefix/?region=us-east-1)
- Support for STORAGE_URL environment variable
- Maintains backward compatibility with individual S3_* env vars
- Auto-detection of S3 from STORAGE_URL (no STORAGE_TYPE needed)
- Support for AWS standard credentials (AWS_ACCESS_KEY_ID/AWS_SECRET_ACCESS_KEY)
- Updated StorageManager to use unified configuration approach
- Updated clean-world.mjs script to support URI configuration
- Added comprehensive .env.example with configuration examples
- Follows AWS credential chain best practices (env vars, profiles, IAM roles)

Examples:
STORAGE_URL=s3://bucket-name/assets/?region=us-east-1
AWS_ACCESS_KEY_ID=AKIA...
AWS_SECRET_ACCESS_KEY=secret...

This addresses the maintainer's request for unified URI-based configuration
similar to DB_URL approach, making S3 configuration cleaner and more intuitive.

DevStarlight committed:
refactor: Simplify configuration to URI-only approach

Remove all legacy individual environment variables and adopt a clean,
consistent URI-based configuration system:

Database Configuration:
- Remove DB_TYPE - auto-detect from DB_URL (postgres:// vs SQLite default)
- Simplified: Only DB_URL needed for PostgreSQL, empty for SQLite

Storage Configuration:
- Remove STORAGE_TYPE, S3_BUCKET_NAME, S3_REGION, S3_*_PREFIX - all in URI now
- Simplified: Only STORAGE_URL needed (s3://bucket/assets/ vs local default)
- Use standard AWS_ACCESS_KEY_ID instead of S3_ACCESS_KEY_ID

Updated Components:
- StorageManager: Only accepts STORAGE_URL, no fallback to individual vars
- clean-world.mjs: Simplified storage/DB config parsing
- db.js: Auto-detect PostgreSQL from postgres:// protocol
- .env.example: Clean and minimal like original, 4 vars instead of 10+

Benefits:
- Consistent with maintainer's request for URI-based config like DB_URL
- Cleaner configuration: one URI per service (DB_URL, STORAGE_URL)
- Auto-detection eliminates need for TYPE variables
- Standard AWS credential naming
- Much simpler .env.example file

Example configuration:
DB_URL=postgres://user:pass@host:port/db
STORAGE_URL=s3://bucket/assets/?region=us-east-1
AWS_ACCESS_KEY_ID=AKIA...
AWS_SECRET_ACCESS_KEY=secret...

DevStarlight committed:
fix: resolve remote storage and database configuration issues

- Fix SSL certificate errors by properly using CloudFront URLs
- Fix PostgreSQL schema configuration with proper search_path
- Fix database migrations to use remote storage instead of local files
- Fix clean-world script to support PostgreSQL schemas
- Remove redundant CloudFront URL configuration
- Improve database configuration structure
